# Feature Selection

## Introduction

You've now learned about many different ways to create features to model more complex relationships. If you included interaction terms between all of your variables and high order polynomial functions, you will no doubt have an issue with overfitting to noise. In this lesson, you'll learn about the different techniques you can use to only use features that are most relevant to your model.

### Objectives

You will be able to:

* List the different methods of feature selection
* Identify when it is appropriate to use certain methods of feature selection
* Use feature selection on a dataset
* Demonstrate how different methods of feature selection are performed

### Purpose of Feature Selection

Feature selection is the process by which you select a subset of features relevant for model construction. Feature Selection comes with several benefits, the most obvious being the improvement of a machine learning algorithm. Other benefits include:

* Decrease in computational complexity: As the number of features are reduced in a model, the easier it will be to compute the parameters of your model. It will also mean a decrease in the amount of data storage required to maintain the features of your model.
* Understanding your data: In the process of feature selection, you will gain more of an understanding of how features relate to one another.

Now, let's look at the different types of feature selection and their advantages/disadvantages.

### Types of Feature Selection

Like many things in data science there is not a clear and easy answer for deciding what features to include in a model. There are however, different strategies you can use to process in the correct manner.


* Domain Knowledge
* Wrapper Methods
* Filter Methods
* Embedded Methods


#### Domain Knowledge   
One of the first steps a data scientist should take when trying to determine what types of features might be useful is knowledge of the specific domain related to your dataset. This might mean reading past research papers that have explored similar topics or asking to key stakeholders to determine what the believe the most important factors are for predicting something.


#### Wrapper Methods   
Wrapper methods determine the optimal subset of features using different combinations of features to train models and then calculating performance. Every subset is used to train models and then evaluated on a test set. As you might imagine, wrapper methods can end up being very computationally intensive, however they are highly effective in determining the optimal subset. Because wrapper methods are so time-consuming, it becomes challenging to use them with large feature sets. 

<img src = "./images/wrapper.png">

An example of a wrapper method in linear regression is Recursive Feature Elimination, which starts with all features included in a model and removes them one by one. After the model has had a feature removed, whichever subset of features resulted in the least statistically significant deterioration of the model fit will indicate which omitted feature is the least useful for prediction. The opposite of this process is Forward Selection, which undergoes the same process in reverse. It begins with a single feature and continues to add the next feature that improves model performance the most. 

#### Filter Methods   
Filter methods are feature selection methods carried out as a pre-processing step before even running a model. Filter methods work by observing characteristics of how variables are related to one another. Depending on the model that is being used, different metrics are used to determine which features will get eliminated and which will remain. Typically, filter methods will return a "feature ranking" that will tell you how features are ordered in relation to one another. They will remove the variables that are considered redundant. It's up to the data scientist to determine the cut-off point at which they will keep the top n features, and this n is usually determined through cross validation.

<img src= "./images/filter.png">

In the linear regression context, a common filter method is to eliminate features that are too highly correlated with one another. Another method is to use a variance threshold. This sets some threshold of required variance for features in order to include them in a model. The thought process behind this is that if variables do not have a high variance, they will not change much and will therefore not have much impact on our dependent variable.


#### Embedded methods   
Embedded methods are feature selection methods that are included within the actual formulation of your machine learning algorithm. The most common kind of embedded method is regularization, in particular Lasso, because it has the capability of reducing your set of features automatically.

<img src = "./images/embedded.png">

## Feature Selection in Action

Now, we're going to review the process behind performing feature selection with a dataset pertaining to diabetes. The dataset contains the independent variables age, sex, body mass index, blood pressure, and 6 different blood serum measurements. The target variable represents a quantitative measurement progression of diabetes from one year after a baseline observation. With feature selection, our goal is to find a model that is able to maintain high accuracy while not overfitting to noise.

### Processing the Data

To begin with, we are going to load the data and create a dummy variable for the variable SEX.

In [1]:
# importing necessary libraries
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from sklearn import datasets, linear_model
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.linear_model import LinearRegression

# reading in the data
df = pd.read_csv('diabetes.tab.txt', sep='\t', lineterminator='\n')


In [3]:
df

AGE  SEX   BMI      BP   S1     S2    S3    S4      S5   S6    Y
0     59    2  32.1  101.00  157   93.2  38.0  4.00  4.8598   87  151
1     48    1  21.6   87.00  183  103.2  70.0  3.00  3.8918   69   75
2     72    2  30.5   93.00  156   93.6  41.0  4.00  4.6728   85  141
3     24    1  25.3   84.00  198  131.4  40.0  5.00  4.8903   89  206
4     50    1  23.0  101.00  192  125.4  52.0  4.00  4.2905   80  135
5     23    1  22.6   89.00  139   64.8  61.0  2.00  4.1897   68   97
6     36    2  22.0   90.00  160   99.6  50.0  3.00  3.9512   82  138
7     66    2  26.2  114.00  255  185.0  56.0  4.55  4.2485   92   63
8     60    2  32.1   83.00  179  119.4  42.0  4.00  4.4773   94  110
9     29    1  30.0   85.00  180   93.4  43.0  4.00  5.3845   88  310
10    22    1  18.6   97.00  114   57.6  46.0  2.00  3.9512   83  101
11    56    2  28.0   85.00  184  144.8  32.0  6.00  3.5835   77   69
12    53    1  23.7   92.00  186  109.2  62.0  3.00  4.3041   81  179
13    50    2  26.2   97.00  186  105.4  49.0  4.00  5.0626   88  185
14    61    1  24.0   91.00  202  115.4  72.0  3.00  4.2905   73  118
15    34    2  24.7  118.00  254  184.2  39.0  7.00  5.0370   81  171
16    47    1  30.3  109.00  207  100.2  70.0  3.00  5.2149   98  166
17    68    2  27.5  111.00  214  147.0  39.0  5.00  4.9416   91  144
18    38    1  25.4   84.00  162  103.0  42.0  4.00  4.4427   87   97
19    41    1  24.7   83.00  187  108.2  60.0  3.00  4.5433   78  168
20    35    1  21.1   82.00  156   87.8  50.0  3.00  4.5109   95   68
21    25    2  24.3   95.00  162   98.6  54.0  3.00  3.8501   87   49
22    25    1  26.0   92.00  187  120.4  56.0  3.00  3.9703   88   68
23    61    2  32.0  103.67  210   85.2  35.0  6.00  6.1070  124  245
24    31    1  29.7   88.00  167  103.4  48.0  4.00  4.3567   78  184
25    30    2  25.2   83.00  178  118.4  34.0  5.00  4.8520   83  202
26    19    1  19.2   87.00  124   54.0  57.0  2.00  4.1744   90  137
27    42    1  31.9   83.00  158   87.6  53.0  3.00  4.4659  101   85
28    63    1  24.4   73.00  160   91.4  48.0  3.00  4.6347   78  131
29    67    2  25.8  113.00  158   54.2  64.0  2.00  5.2933  104  283
..   ...  ...   ...     ...  ...    ...   ...   ...     ...  ...  ...
412   69    1  34.3  113.00  200  123.8  54.0  4.00  4.7095  112  261
413   34    1  26.3   87.00  197  120.0  63.0  3.00  4.2485   96  113
414   71    2  27.0   93.33  269  190.2  41.0  6.56  5.2417   93  131
415   47    1  27.2   80.00  208  145.6  38.0  6.00  4.8040   92  174
416   41    1  33.8  123.33  187  127.0  45.0  4.16  4.3175  100  257
417   34    1  33.0   73.00  178  114.6  51.0  3.49  4.1271   92   55
418   51    1  24.1   87.00  261  175.6  69.0  4.00  4.4067   93   84
419   43    1  21.3   79.00  141   78.8  53.0  3.00  3.8286   90   42
420   55    1  23.0   94.67  190  137.6  38.0  5.00  4.2767  106  146
421   59    2  27.9  101.00  218  144.2  38.0  6.00  5.1874   95  212
422   27    2  33.6  110.00  246  156.6  57.0  4.00  5.0876   89  233
423   51    2  22.7  103.00  217  162.4  30.0  7.00  4.8122   80   91
424   49    2  27.4   89.00  177  113.0  37.0  5.00  4.9053   97  111
425   27    1  22.6   71.00  116   43.4  56.0  2.00  4.4188   79  152
426   57    2  23.2  107.33  231  159.4  41.0  5.63  5.0304  112  120
427   39    2  26.9   93.00  136   75.4  48.0  3.00  4.1431   99   67
428   62    2  34.6  120.00  215  129.2  43.0  5.00  5.3660  123  310
429   37    1  23.3   88.00  223  142.0  65.0  3.40  4.3567   82   94
430   46    1  21.1   80.00  205  144.4  42.0  5.00  4.5326   87  183
431   68    2  23.5  101.00  162   85.4  59.0  3.00  4.4773   91   66
432   51    1  31.5   93.00  231  144.0  49.0  4.70  5.2523  117  173
433   41    1  20.8   86.00  223  128.2  83.0  3.00  4.0775   89   72
434   53    1  26.5   97.00  193  122.4  58.0  3.00  4.1431   99   49
435   45    1  24.2   83.00  177  118.4  45.0  4.00  4.2195   82   64
436   33    1  19.5   80.00  171   85.4  75.0  2.00  3.9703   80   48
437   60    2  

In [2]:
# obtaining the X and Y variables from the dataframe
features = df.iloc[:,:-1]
target = df['Y']
# creating dummy variable for sex
features['female'] = pd.get_dummies(features['SEX'],drop_first=True)
features.drop(columns=['SEX'], inplace=True)
features.head()

AGE   BMI     BP   S1     S2    S3   S4      S5  S6  female
0   59  32.1  101.0  157   93.2  38.0  4.0  4.8598  87       1
1   48  21.6   87.0  183  103.2  70.0  3.0  3.8918  69       0
2   72  30.5   93.0  156   93.6  41.0  4.0  4.6728  85       1
3   24  25.3   84.0  198  131.4  40.0  5.0  4.8903  89       0
4   50  23.0  101.0  192  125.4  52.0  4.0  4.2905  80       0

For both regularization (an embedded method) and various filters, it is important to standardize the data. This next cell is fitting the data to a StandardScaler from sklearn.

In [4]:
from sklearn import preprocessing
X_train, X_test, y_train, y_test = train_test_split(features, target, random_state=20,test_size=0.2)
scaler = preprocessing.StandardScaler()

## scaling every feature except the binary column female
scaler.fit(X_train.iloc[:,:-1])
transformed_training_features = scaler.transform(X_train.iloc[:,:-1])
transformed_testing_features = scaler.transform(X_test.iloc[:,:-1])

X_train_transformed = pd.DataFrame(scaler.transform(X_train.iloc[:,:-1]), columns=X_train.columns[:-1], index=X_train.index)
X_train_transformed['female'] = X_train['female']

X_test_transformed = pd.DataFrame(scaler.transform(X_test.iloc[:,:-1]), columns=X_train.columns[:-1], index=X_test.index)
X_test_transformed['female'] = X_test['female']

/opt/conda/envs/learn-env/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/opt/conda/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:7: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  import sys
/opt/conda/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:8: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  
/opt/conda/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:10: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  # Remove the CWD from sys.path while we load stuff.
/opt/conda/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:13: DataConversionWarning: Data with input dtype int64, flo

In [7]:
X_train

AGE   BMI      BP   S1     S2    S3    S4      S5   S6  female
171   43  22.1   66.00  134   77.2  45.0  3.00  4.0775   80       0
388   63  24.1  111.00  184  112.2  44.0  4.00  4.9345   82       1
268   66  34.6  115.00  204  139.4  36.0  6.00  4.9628  109       1
31    42  20.3   71.00  161   81.2  66.0  2.00  4.2341   81       0
427   39  26.9   93.00  136   75.4  48.0  3.00  4.1431   99       1
311   75  30.1   78.00  222  154.2  44.0  5.05  4.7791   97       0
374   19  23.2   75.00  143   70.4  52.0  3.00  4.6347   72       0
185   28  24.7   97.00  175   99.6  32.0  5.00  5.3799   87       1
81    52  24.3   86.00  197  133.6  44.0  5.00  4.5747   91       1
45    56  23.1  104.00  181  116.4  47.0  4.00  4.4773   79       1
413   34  26.3   87.00  197  120.0  63.0  3.00  4.2485   96       0
332   57  36.1  117.00  181  108.2  34.0  5.00  5.2679  100       0
28    63  24.4   73.00  160   91.4  48.0  3.00  4.6347   78       0
129   51  27.9   96.00  196  122.2  42.0  5.00  5.0689  120       0
200   64  21.0   92.33  227  146.8  65.0  3.49  4.3307  102       0
342   64  28.4  111.00  184  127.0  41.0  4.00  4.3820   97       1
433   41  20.8   86.00  223  128.2  83.0  3.00  4.0775   89       0
165   37  20.2   81.00  162   87.8  63.0  3.00  4.0254   88       0
197   62  26.7  115.00  183  124.0  35.0  5.00  4.7875  100       1
115   40  26.5   93.00  236  147.0  37.0  7.00  5.5607   92       1
98    49  25.9   91.00  178  106.6  52.0  3.00  4.5747   75       1
418   51  24.1   87.00  261  175.6  69.0  4.00  4.4067   93       0
189   48  20.2   95.00  187  117.4  53.0  4.00  4.4188   85       0
417   34  33.0   73.00  178  114.6  51.0  3.49  4.1271   92       0
89    36  21.9   89.00  189  105.2  68.0  3.00  4.3694   96       0
117   65  24.4  120.00  222  135.6  37.0  6.00  5.5094  124       0
121   66  28.0  101.00  195  129.2  40.0  5.00  4.8598   94       0
0     59  32.1  101.00  157   93.2  38.0  4.00  4.8598   87       1
21    25  24.3   95.00  162   98.6  54.0  3.00  3.8501   87       1
429   37  23.3   88.00  223  142.0  65.0  3.40  4.3567   82       0
..   ...   ...     ...  ...    ...   ...   ...     ...  ...     ...
387   36  27.8   73.00  153  104.4  42.0  4.00  3.4965   73       0
25    30  25.2   83.00  178  118.4  34.0  5.00  4.8520   83       1
141   41  30.8   81.00  214  152.0  28.0  7.60  5.1358  123       0
428   62  34.6  120.00  215  129.2  43.0  5.00  5.3660  123       1
6     36  22.0   90.00  160   99.6  50.0  3.00  3.9512   82       1
255   49  20.3   93.00  184  103.0  61.0  3.00  4.6052   93       0
263   28  19.2   81.00  155   94.6  51.0  3.00  3.8501   87       1
357   32  31.0  100.00  177   96.2  45.0  4.00  5.1874   77       1
105   55  23.4   83.00  166  101.6  46.0  4.00  4.5218   96       1
144   57  30.7   90.00  204  147.8  34.0  6.00  4.7095   93       1
190   51  25.2  103.00  176  112.2  37.0  5.00  4.8978   90       0
424   49  27.4   89.00  177  113.0  37.0  5.00  4.9053   97       1
230   38  33.0   78.00  301  215.0  50.0  6.02  5.1930  108       1
253   71  29.5   97.00  227  151.6  45.0  5.00  5.0239  108       0
272   52  27.0   78.33  134   73.0  44.0  3.05  4.4427   69       0
251   34  35.5  120.00  233  146.6  34.0  7.00  5.5683  101       1
282   68  25.9   93.00  253  181.2  53.0  5.00  4.5433   98       0
40    50  25.6  101.00  229  162.2  43.0  5.00  4.7791  114       1
118   33  25.4  102.00  206  141.0  39.0  5.00  4.8675  105       1
162   34  29.2   73.00  172  108.2  49.0  4.00  4.3041   91       0
71    48  29.5  131.00  207  132.2  47.0  4.00  4.9345  106       0
278   67  23.0   70.00  184  128.0  35.0  5.00  4.6540   99       1
331   71  24.0   84.00  138   85.8  39.0  4.00  4.1897   90       1
404   44  31.4  115.00  165   97.6  52.0  3.00  4.3438   89       0
393   28  22.1   82.00  168  100.6  54.0  3.00  4.2047   86       0
218   37  20.3   83.00  185  124.6  38.0  5.00  4.7185   88       0
412   69  34.3  113.00  200  123.8  54.0  4.00  4.7095  112  

In [6]:
X_train_transformed

AGE       BMI        BP        S1        S2        S3        S4  \
171 -0.433522 -0.967597 -2.067847 -1.623215 -1.280312 -0.347527 -0.852832   
388  1.117754 -0.516691  1.142458 -0.168101 -0.129601 -0.424950 -0.083651   
268  1.350445  1.850570  1.427819  0.413945  0.764667 -1.044334  1.454710   
31  -0.511086 -1.373413 -1.711146 -0.837453 -1.148802  1.278358 -1.622013   
427 -0.743778  0.114579 -0.141664 -1.565010 -1.339491 -0.115257 -0.852832   
311  2.048520  0.836030 -1.211765  0.937786  1.251253 -0.424950  0.723989   
374 -2.295054 -0.719599 -1.425786 -1.361294 -1.503879  0.194435 -0.852832   
185 -1.596980 -0.381419  0.143697 -0.430021 -0.543857 -1.354027  0.685530   
81   0.264552 -0.471600 -0.641044  0.210229  0.573977 -0.424950  0.685530   
45   0.574807 -0.742144  0.643077 -0.255408  0.008485 -0.192680 -0.083651   
413 -1.131597 -0.020693 -0.569704  0.210229  0.126844  1.046089 -0.852832   
332  0.652371  2.188750  1.570499 -0.255408 -0.261110 -1.199180  0.685530   
28   1.117754 -0.449055 -1.568466 -0.866555 -0.813452 -0.115257 -0.852832   
129  0.186988  0.340032  0.072357  0.181127  0.199174 -0.579796  0.685530   
200  1.195318 -1.215596 -0.189462  1.083297  1.007960  1.200935 -0.475933   
342  1.195318  0.452759  1.142458 -0.168101  0.356986 -0.657219 -0.083651   
433 -0.588650 -1.260687 -0.641044  0.966888  0.396439  2.594550 -0.852832   
165 -0.898905 -1.395959 -0.997745 -0.808351 -0.931811  1.046089 -0.852832   
197  1.040190  0.069488  1.427819 -0.197203  0.258354 -1.121757  0.685530   
115 -0.666214  0.024397 -0.141664  1.345218  1.014535 -0.966911  2.223891   
98   0.031860 -0.110875 -0.284344 -0.342714 -0.313714  0.194435 -0.852832   
418  0.186988 -0.516691 -0.569704  2.072775  1.954831  1.510627 -0.083651   
189 -0.045703 -1.395959  0.001017 -0.080794  0.041362  0.271858 -0.083651   
417 -1.131597  1.489844 -1.568466 -0.342714 -0.050695  0.117012 -0.475933   
89  -0.976469 -1.012688 -0.427024 -0.022589 -0.359743  1.433204 -0.852832   
117  1.272882 -0.449055  1.784519  0.937786  0.639732 -0.966911  1.454710   
121  1.350445  0.362578  0.429057  0.152024  0.429316 -0.734642  0.685530   
0    0.807499  1.286936  0.429057 -0.953862 -0.754272 -0.889488 -0.083651   
21  -1.829671 -0.471600  0.001017 -0.808351 -0.576734  0.349281 -0.852832   
429 -0.898905 -0.697053 -0.498364  0.966888  0.850148  1.200935 -0.545160   
..        ...       ...       ...       ...       ...       ...       ...   
387 -0.976469  0.317487 -1.568466 -1.070271 -0.386045 -0.579796 -0.083651   
25  -1.441852 -0.268692 -0.855065 -0.342714  0.074240 -1.199180  0.685530   
141 -0.588650  0.993847 -0.997745  0.704968  1.178923 -1.663719  2.685399   
428  1.040190  1.850570  1.784519  0.734070  0.429316 -0.502373  0.685530   
6   -0.976469 -0.990143 -0.355684 -0.866555 -0.543857  0.039589 -0.852832   
255  0.031860 -1.373413 -0.141664 -0.168101 -0.432073  0.891243 -0.852832   
263 -1.596980 -1.621412 -0.997745 -1.012067 -0.708244  0.117012 -0.852832   
357 -1.286724  1.038938  0.357717 -0.371817 -0.655640 -0.347527 -0.083651   
105  0.497243 -0.674508 -0.855065 -0.691942 -0.478102 -0.270103 -0.083651   
144  0.652371  0.971302 -0.355684  0.413945  1.040837 -1.199180  1.454710   
190  0.186988 -0.268692  0.571737 -0.400919 -0.129601 -0.966911  0.685530   
424  0.031860  0.227306 -0.427024 -0.371817 -0.103299 -0.966911  0.685530   
230 -0.821342  1.489844 -1.211765  3.236866  3.250203  0.039589  1.470094   
253  1.738264  0.700758  0.143697  1.083297  1.165772 -0.347527  0.685530   
272  0.264552  0.137124 -1.188223 -1.623215 -1.418397 -0.424950 -0.814373   
251 -1.131597  2.053478  1.784519  1.257911  1.001384 -1.199180  2.223891   
282  1.505573 -0.110875 -0.141664  1.839956  2.138945  0.271858  0.685530   
40   0.109424 -0.178511  0.429057  1.141502  1.514273 -0.502373  0.685530   
118 -1.209161 -0.223601  0.500397  0.472149  0.817270 -0.812065  0.685530   
162 -1.131597  0.633122 -1.568466 -0.517328 -0.261110 -0.037834 -0.083651   
71  -0

Before we start subsetting features, we should see how well the model performs without performing any kind of interactions added or feature selection performed. Because we are going to be running many different models here, we have created a function to ensure that we are remaining D.R.Y. with our code.

In [8]:
def run_model(model,X_train,X_test,y_train,y_test):
    
    print('Training R^2 :',model.score(X_train,y_train))
    y_pred_train = model.predict(X_train)
    print('Training Root Mean Square Error',np.sqrt(metrics.mean_squared_error(y_train,y_pred_train)))
    print('\n----------------\n')
    print('Testing R^2 :',model.score(X_test,y_test))
    y_pred_test = model.predict(X_test)
    print('Testing Root Mean Square Error',np.sqrt(metrics.mean_squared_error(y_test,y_pred_test)))

In [9]:
lm = LinearRegression()
lm.fit(X_train_transformed,y_train)
run_model(lm,X_train_transformed,X_test_transformed,y_train,y_test)

Training R^2 : 0.5371947100976313
Training Root Mean Square Error 52.21977472848369

----------------

Testing R^2 : 0.4179775463198646
Testing Root Mean Square Error 58.835897088895045


The model has not performed exceptionally well here, so we can try adding some additional features. Let's go ahead and add a polynomial degree of up to 3.

In [10]:
poly = preprocessing.PolynomialFeatures(degree=2, interaction_only=False, include_bias=False)
X_poly_train = pd.DataFrame(poly.fit_transform(X_train_transformed), columns=poly.get_feature_names(features.columns))
X_poly_test = pd.DataFrame(poly.transform(X_test_transformed), columns=poly.get_feature_names(features.columns))
X_poly_train.head()

AGE       BMI        BP        S1        S2        S3        S4  \
0 -0.433522 -0.967597 -2.067847 -1.623215 -1.280312 -0.347527 -0.852832   
1  1.117754 -0.516691  1.142458 -0.168101 -0.129601 -0.424950 -0.083651   
2  1.350445  1.850570  1.427819  0.413945  0.764667 -1.044334  1.454710   
3 -0.511086 -1.373413 -1.711146 -0.837453 -1.148802  1.278358 -1.622013   
4 -0.743778  0.114579 -0.141664 -1.565010 -1.339491 -0.115257 -0.852832   

         S5        S6  female    ...         S4^2     S4 S5     S4 S6  \
0 -1.095555 -1.006077     0.0    ...     0.727322  0.934324  0.858015   
1  0.543382 -0.831901     1.0    ...     0.006998 -0.045455  0.069589   
2  0.597504  1.519478     1.0    ...     2.116182  0.869195  2.210400   
3 -0.796071 -0.918989     0.0    ...     2.630925  1.291237  1.490612   
4 -0.970101  0.648597     1.0    ...     0.727322  0.827333 -0.553144   

   S4 female      S5^2     S5 S6  S5 female      S6^2  S6 female  female^2  
0  -0.000000  1.200240  1.102213  -0.000000  1.012192  -0.000000       0.0  
1  -0.083651  0.295264 -0.452040   0.543382  0.692060  -0.831901       1.0  
2   1.454710  0.357011  0.907894   0.597504  2.308813   1.519478       1.0  
3  -0.000000  0.633729  0.731581  -0.000000  0.844541  -0.000000       0.0  
4  -0.852832  0.941095 -0.629204  -0.970101  0.420678   0.648597       1.0  

[5 rows x 65 columns]

As you can see, this has now created 65 total columns! You can imagine that this model will greatly overfit to the data. Let's try it out with our training and test set.

In [11]:
lr_poly = LinearRegression()
lr_poly.fit(X_poly_train,y_train)

run_model(lr_poly,X_poly_train,X_poly_test,y_train,y_test)

Training R^2 : 0.623742432676387
Training Root Mean Square Error 47.08455372329227

----------------

Testing R^2 : 0.3688694444251702
Testing Root Mean Square Error 61.267775626915494


Clearly, the model has fit very well to the training data, but it has fit to a lot of noise. The testing $R^{2}$ is worse than our simple model we fit previously! It's time to get rid of some features to see if this improves the model.

## Filter Methods  
Let's begin by trying out some filter methods for feature selection. The benefit of filter methods is that they can provide us with some useful visualizations for helping us gain an understanding about characteristics of our data. To begin with, let's use a simple variance threshold to eliminate those features with a low variance.

In [12]:
from sklearn.feature_selection import VarianceThreshold

threshold_ranges = np.linspace(0,2,num=6)

for thresh in threshold_ranges:
    print(thresh)
    selector = VarianceThreshold(thresh)
    reduced_feature_train = selector.fit_transform(X_poly_train)
    reduced_feature_test = selector.transform(X_poly_test)
    lr = LinearRegression()
    lr.fit(reduced_feature_train,y_train)
    run_model(lr,reduced_feature_train,reduced_feature_test,y_train,y_test)
    print('--------------------------------------------------------------------')

0.0
Training R^2 : 0.6237424326763871
Training Root Mean Square Error 47.08455372329226

----------------

Testing R^2 : 0.36886944442517056
Testing Root Mean Square Error 61.26777562691547
--------------------------------------------------------------------
0.4
Training R^2 : 0.6035018897144957
Training Root Mean Square Error 48.33440733222434

----------------

Testing R^2 : 0.3580801962858088
Testing Root Mean Square Error 61.78924619409419
--------------------------------------------------------------------
0.8
Training R^2 : 0.5894227238666293
Training Root Mean Square Error 49.18506972762005

----------------

Testing R^2 : 0.3640169603035057
Testing Root Mean Square Error 61.50285507146046
--------------------------------------------------------------------
1.2000000000000002
Training R^2 : 0.1991536009695497
Training Root Mean Square Error 68.69267841228015

----------------

Testing R^2 : 0.036254917874099735
Testing Root Mean Square Error 75.71006122152009
-------------------

Hmmm, that did not seem to eliminate the features very well. It's better than the base polynomial, but it is not any better than some of the more complicated.

In [13]:
from sklearn.feature_selection import f_regression, mutual_info_regression, SelectKBest
selector = SelectKBest(score_func=f_regression)
X_k_best_train = selector.fit_transform(X_poly_train, y_train)
X_k_best_test= selector.transform(X_poly_test)
lr = LinearRegression()
lr.fit(X_k_best_train ,y_train)
run_model(lr,X_k_best_train,X_k_best_test,y_train,y_test)

Training R^2 : 0.5229185029521006
Training Root Mean Square Error 53.01907218972858

----------------

Testing R^2 : 0.42499888052723567
Testing Root Mean Square Error 58.4799314703427


In [14]:
selector = SelectKBest(score_func=mutual_info_regression)
X_k_best_train = selector.fit_transform(X_poly_train, y_train)
X_k_best_test= selector.transform(X_poly_test)
lr = LinearRegression()
lr.fit(X_k_best_train ,y_train)
run_model(lr,X_k_best_train,X_k_best_test,y_train,y_test)

/opt/conda/envs/learn-env/lib/python3.6/site-packages/sklearn/utils/validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by the scale function.
  warnings.warn(msg, DataConversionWarning)


Training R^2 : 0.48788560532499714
Training Root Mean Square Error 54.93123060353359

----------------

Testing R^2 : 0.39166279490049793
Testing Root Mean Square Error 60.151255840313844


## Wrapper Methods

Now let's use Recursive Feature elimination (RFE) to try out a wrapper method. You'll notice that sci-kit learn has a built in RFECV function, which automatically determines the optimal number of features to keep when it is run based off the estimator that is passed into it. Here it is in action: 

In [15]:
from sklearn.feature_selection import RFE, RFECV
from sklearn.linear_model import LinearRegression

rfe = RFECV(LinearRegression(),cv=5)
X_rfe_train = rfe.fit_transform(X_poly_train,y_train)
X_rfe_test = rfe.transform(X_poly_test)
lm = LinearRegression().fit(X_rfe_train,y_train)
run_model(lm,X_rfe_train,X_rfe_test,y_train,y_test)
print ("The optimal number of features is: ",rfe.n_features_)

Training R^2 : 0.37930327292162724
Training Root Mean Square Error 60.474956480110215

----------------

Testing R^2 : 0.37417761375281045
Testing Root Mean Square Error 61.0095830577443
The optimal number of features is:  10


With Recursive Feature Elimination, we went from an $R^2$ score of 0.368 to 0.374 (a tiny bit better). Let's see if we can improve upon these results even more by trying embedded methods.

## Embedded Methods  
To compare to our other methods, we will use Lasso as the embedded method of feature selection. Luckily for us, sklearn has a builtin method to help us find the optimal features! It performs cross validation to determine the correct regularization parameter (how much to penalize our function).

In [16]:
from sklearn.linear_model import LassoCV
lasso=LassoCV(max_iter=100000,cv=5)
lasso.fit(X_train_transformed,y_train)
run_model(lasso,X_train_transformed,X_test_transformed,y_train,y_test)
print("The optimal alpha for the Lasso Regression is: ",lasso.alpha_)

Training R^2 : 0.535154465585244
Training Root Mean Square Error 52.33475174961373

----------------

Testing R^2 : 0.4267044232634858
Testing Root Mean Square Error 58.39313677555575
The optimal alpha for the Lasso Regression is:  0.23254844944953365


Let's compare this to a model with all of the polynomial features included.

In [17]:
lasso2 = LassoCV(max_iter=100000,cv=5)

lasso2.fit(X_poly_train,y_train)
run_model(lasso2,X_poly_train,X_poly_test,y_train,y_test)
print("The optimal alpha for the Lasso Regression is: ",lasso2.alpha_)

Training R^2 : 0.5635320505309955
Training Root Mean Square Error 50.71214913665364

----------------

Testing R^2 : 0.43065954589620015
Testing Root Mean Square Error 58.191363260874226
The optimal alpha for the Lasso Regression is:  1.7591437388826359


As we can see, the regularization had minimal effect on the performance of the model, but it did improve the RMSE for the test set ever so slightly! There are no set steps someone should take in order to determine the optimal feature set. In fact, now there are automated machine learning pipelines that will determine the optimal subset of features for a given problem. One of the most important and often overlooked methods of feature selection is using domain knowledge about a given area to either eliminate features or create new ones.

__Additional Resources__: 

[Feature Selection](https://www.researchgate.net/profile/Amparo_Alonso-Betanzos/publication/221252792_Filter_Methods_for_Feature_Selection_-_A_Comparative_Study/links/543fd9ec0cf21227a11b8e05.pdf)


[An Introduction to Variable and Feature Selection](http://www.jmlr.org/papers/volume3/guyon03a/guyon03a.pdf)

## Summary

This lesson formalized the different types of feature selection and introduced some new techniques to you. You learned about Filter Methods, Wrapper Methods, and Embedded Methods as well as advantages and disadvantages to both. In the next lab, you'll get a chance to test out these feature selection methods and everything else you've learned in this section to make the optimal model!